### PR TITLE
PS-9048 fix 8.0: Fixed problem with percent character in n-grams

### PR DESCRIPTION
**Pull request by percona-ysorokin**
https://perconadev.atlassian.net/browse/PS-9048

Fixed problem with 'fts_index_fetch_nodes()' function not being able to properly handle situations when 'word' parameter contained special characters used by 'LIKE' clauses ('%' and '_'). Introduced additional boolean parameter 'exact_match' that instructs this function to use either 'WHERE word = :word' or 'WHERE word LIKE :word' clauses when selecting records from internal FTS tables. We call 'fts_index_fetch_nodes()' with 'exact_match' set to 'true' only from 'fts_optimize_words()' (when we perform 'OPTIMIZE TABLE' under 'innodb_optimize_fulltext_only' enabled). In every other place
* fts_query_difference()
* fts_query_intersect()
* fts_query_union()
* fts_query_phrase_search() where we indeed need pattern matching we call this function with 'exact_match' set to 'false' (instructing the function to use the 'LIKE' clause).

Added two new MTR test cases:
* 'innodb_fts.percona_ft_ngram_special_chars_ewc_on'
* 'innodb_fts.percona_ft_ngram_special_chars_ewc_off' which reproduce original crash scenario under both 'ft_query_extra_word_chars' set to 'ON' and 'OFF'. These tests cases also check ngram parsing / querying strings containing various special characters. Please note that they have "result mismatch" status strings recorded in the '.result' files for certain combinations of special characters and this is expected behavior.

### CODE REVIEW

**Review by oleksandr-kachan — APPROVED** [review]

**Comment by percona-ysorokin** [comment]
Jenkins build link
https://ps80.cd.percona.com/view/8.0%20parallel%20MTR/job/percona-server-8.0-param-parallel-mtr/36/

File: mysql-test/suite/innodb_fts/t/percona_ft_special_chars_mecab_ewc_on.test [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#
+# PS-9048: Debug assertion with OPTIMIZE table and fulltext indexes in InnoDB
+# https://perconadev.atlassian.net/browse/PS-9048
+#
+--source include/have_ngram.inc
```
Comment (oleksandr-kachan) [inline mysql-test/suite/innodb_fts/t/percona_ft_special_chars_mecab_ewc_on.test]: have_mecab

File: mysql-test/suite/innodb_fts/t/percona_ft_special_chars_mecab_ewc_on.test [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+--source suite/innodb_fts/include/percona_install_mecab_plugin.inc
+
+SET ft_query_extra_word_chars = ON;
+
+--let $fts_parser =
```
Comment (oleksandr-kachan) [inline mysql-test/suite/innodb_fts/t/percona_ft_special_chars_mecab_ewc_on.test]: Shall it be set to 'mecab' directly?

File: mysql-test/suite/innodb_fts/t/percona_ft_special_chars_mecab_ewc_off.test [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#
+# PS-9048: Debug assertion with OPTIMIZE table and fulltext indexes in InnoDB
+# https://perconadev.atlassian.net/browse/PS-9048
+#
+--source include/have_ngram.inc
```
Comment (oleksandr-kachan) [inline mysql-test/suite/innodb_fts/t/percona_ft_special_chars_mecab_ewc_off.test]: have_mecab

**Review by oleksandr-kachan — CHANGES_REQUESTED** [review]
Please check comments in code

File: mysql-test/suite/innodb_fts/t/percona_ft_special_chars_mecab_ewc_on.test (hunk 1 above)
Comment (percona-ysorokin) [reply mysql-test/suite/innodb_fts/t/percona_ft_special_chars_mecab_ewc_on.test]: fixed

File: mysql-test/suite/innodb_fts/t/percona_ft_special_chars_mecab_ewc_on.test (hunk 2 above)
Comment (percona-ysorokin) [reply mysql-test/suite/innodb_fts/t/percona_ft_special_chars_mecab_ewc_on.test]: fixed

File: mysql-test/suite/innodb_fts/t/percona_ft_special_chars_mecab_ewc_off.test (hunk 3 above)
Comment (percona-ysorokin) [reply mysql-test/suite/innodb_fts/t/percona_ft_special_chars_mecab_ewc_off.test]: fixed

**Review by oleksandr-kachan — APPROVED** [review]

**Review by oleksandr-kachan — APPROVED** [review]

**Pull request merged**